### PR TITLE
[GPU] NCCL User Buffer registration (1/3)

### PR DESCRIPTION
**Pull request by trevor-m**
This PR enables XLA to take advantage of [NCCL User Buffer Registration](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/bufferreg.html) in NCCL 2.19:
> User Buffer Registration is a feature that allows NCCL to directly send/receive/operate data through the user buffer without extra internal copy. It can accelerate collectives and reduce the resource usage (e.g. #channel usage).

NCCL supports this feature for all reduce, all gather, and reduce scatter. To use the feature, A buffer must be allocated with `ncclMemAlloc`. Then, the buffer must be registered using `ncclCommRegister` for all communicators. If capturing collectives using cuda graphs, `ncclCommRegister` can be skipped. Finally, when performing a collective using offsets to this buffer, all devices must have the same offset into the buffer.

This PR contains the following components to enable this feature:

1. Compiler
   - If flag `xla_gpu_enable_nccl_user_buffers` is true, a custom "colorer" function will be passed to `BufferAssignment` which will mark all buffers for all reduce, all gather, and reduce scatter to use alternate memory space `1`.
2. Runtime
   -  The StreamExecutorGpuClient will create a second `BFCAllocator` for each device which uses `ncclMemAlloc/ncclMemFree` as its suballocator. The amount of memory reserved can be configured using `GpuAllocatorConfig.collective_memory_size`.
   - `MultiDeviceAllocator` will read the `memory_space` value for each allocation and use it to route the allocation to the correct allocator. Since `memory_space` is not passed to the deallocate function, it uses a hash map to remember which memory space each address belongs to.
   - When running a collective, the first time a buffer is encountered for each communicator it is registered using `ncclCommRegister`. The collective buffers will use a slice of the `ncclMemAlloc`'d CollectiveBFCAllocator memory, but `ncclCommRegister` will automatically locate the base address and size to register the entire `ncclMemAlloc` memory. Once, cudagraph capture of nccl collectives is supported, this code can be removed.

NCCL 2.19 is required for `ncclCommRegister`.

PJRT Changes: https://github.com/openxla/xla/pull/7963
StreamExecutor Changes: https://github.com/openxla/xla/pull/7962

### CODE REVIEW

File: xla/pjrt/c/pjrt_c_api_gpu_internal.cc [hunk 1]
```diff
@@ -96,6 +98,10 @@ PJRT_Error* PJRT_Client_Create(PJRT_Client_Create_Args* args) {
       it != create_options.end()) {
     allocator_config.preallocate = std::get<bool>(it->second);
   }
+  if (auto it = create_options.find("collective_memory_size");
+      it != create_options.end()) {
+    allocator_config.collective_memory_size = std::get<int64_t>(it->second);
```
Comment (cheshire) [inline xla/pjrt/c/pjrt_c_api_gpu_internal.cc:103]: Do you think we could move out PJRT changes into a separate PR?

File: xla/service/gpu/compile_module_to_llvm_ir.cc [hunk 2]
```diff
@@ -298,6 +298,34 @@ StatusOr<CompileModuleResults> CompileModuleToLlvmIr(
...
+      for (const auto& alias : buffer.values()) {
+        if (alias->instruction()->opcode() == HloOpcode::kAllReduce ||
+            alias->instruction()->opcode() == HloOpcode::kAllReduceStart ||
+            alias->instruction()->opcode() == HloOpcode::kAllReduceDone ||
+            alias->instruction()->opcode() == HloOpcode::kAllGather ||
+            alias->instruction()->opcode() == HloOpcode::kAllGatherStart ||
+            alias->instruction()->opcode() == HloOpcode::kAllGatherDone ||
+            alias->instruction()->opcode() == HloOpcode::kReduceScatter) {
+          value->set_color(1);
+        } else if ((alias->instruction()->opcode() == HloOpcode::kAsyncStart ||
```
Comment (cheshire) [inline xla/service/gpu/compile_module_to_llvm_ir.cc:315]: Nit: why the separation?

File: xla/service/gpu/gpu_executable.cc [hunk 3]
```diff
@@ -418,7 +418,9 @@ StatusOr<se::DeviceMemoryBase> GpuExecutable::BufferForAllocation(
     se::DeviceMemoryBase buffer_address;
     if (buffer_size > 0) {
       StatusOr<se::OwningDeviceMemory> buffer =
-          memory_allocator->Allocate(device_ordinal, buffer_size);
+          memory_allocator->Allocate(device_ordinal, buffer_size,
```
Comment (cheshire) [inline xla/service/gpu/gpu_executable.cc:421]: Probably can be a separate NFC?

Comment (cheshire) [reply xla/service/gpu/gpu_executable.cc:421]: Same below.

File: xla/service/gpu/gpu_executable.cc [hunk 4]
```diff
@@ -825,7 +829,10 @@ Status GpuExecutable::SetUpMlirAllocation(
     absl::flat_hash_map<ShapeIndex, GpuExecutable::OutputInfo>* output_info,
     Shape* output_shape) {
   for (int i = 0; i < buffer_sizes.size(); i++) {
-    allocations->emplace_back(i, buffer_sizes[i], 0);
+    // This code path is taken when using the non-thunk based runtime. Memory
+    // space is being set to 0 for all allocations. We need to copy over the
+    // value from BufferAssignment instead.
+    allocations->emplace_back(i, buffer_sizes[i], /*memory_space=*/0);
```
Comment (cheshire) [inline xla/service/gpu/gpu_executable.cc:835]: Could you clarify? This feature is only working for the thunk runtime?

File: xla/service/gpu/nccl_collective_thunk.cc [hunk 5]
```diff
@@ -224,6 +224,49 @@ StatusOr<std::vector<DeviceBufferPair>> ConvertToDeviceBuffers(
   return device_buffers;
 }
 
+Status MaybeRegisterBuffers(int device_ordinal,
+                            const std::vector<DeviceBufferPair>& buffers,
+                            ncclComm_t comm) {
+  // TODO(tmorris): Only register if xla_gpu_enable_nccl_user_buffers is true.
```
Comment (cheshire) [inline xla/service/gpu/nccl_collective_thunk.cc:230]: If we don't, what happens?

File: xla/service/gpu/nccl_collective_thunk.cc [hunk 6]
```diff
@@ -224,6 +224,49 @@ StatusOr<std::vector<DeviceBufferPair>> ConvertToDeviceBuffers(
...
+#if XLA_ENABLE_XCCL
+  // Keep track of which communicators we have registered for already.
+  // Each device has one NCCL buffer which only needs to be registered once per
+  // each comm.
+  struct RegisteredBuffers {
```
Comment (cheshire) [inline xla/service/gpu/nccl_collective_thunk.cc:235]: Nit: probably more readable to have the struct outside, in an anonymous namespace.

File: xla/stream_executor/cuda/cuda_driver.cc [hunk 7]
```diff
@@ -56,6 +52,11 @@ limitations under the License.
 #include "tsl/platform/status.h"
 #include "tsl/platform/statusor.h"
 #include "tsl/platform/threadpool.h"
+#include "xla/stream_executor/device_options.h"
```
Comment (cheshire) [inline xla/stream_executor/cuda/cuda_driver.cc:55]: Sorry please avoid reordering for now.

File: xla/stream_executor/integrations/device_mem_allocator.h [hunk 8]
```diff
@@ -52,8 +54,10 @@ class DeviceMemAllocator : public tsl::SubAllocator {
     void* ptr = nullptr;
     *bytes_received = num_bytes;
     if (num_bytes > 0) {
-      if (use_unified_memory_) {
+      if (memory_type_ == MemoryType::kUnified) {
```
Comment (cheshire) [inline xla/stream_executor/integrations/device_mem_allocator.h:57]: Refactoring unified memory handling could be a separate NFC?

File: xla/stream_executor/integrations/tf_allocator_adapter.cc [hunk 9]
```diff
@@ -37,7 +37,9 @@ TfAllocatorAdapter::~TfAllocatorAdapter() {}
 tsl::StatusOr<OwningDeviceMemory> TfAllocatorAdapter::Allocate(
     int device_ordinal, uint64_t size, bool retry_on_failure,
     int64_t memory_space) {
-  CHECK_EQ(memory_space, 0);
+  // Memory space is unused here as it has already been used to select a
+  // different TfAllocatorAdapter.
+  // CHECK_EQ(memory_space, 0);
```
Comment (cheshire) [inline xla/stream_executor/integrations/tf_allocator_adapter.cc:42]: Remove then?

File: xla/stream_executor/integrations/tf_allocator_adapter.h [hunk 10]
```diff
@@ -137,15 +157,25 @@ class MultiDeviceAdapter : public DeviceMemoryAllocator {
...
   }
 
  private:
-  std::vector<std::unique_ptr<TfAllocatorAdapter>> per_device_allocators_;
+  absl::flat_hash_map<int64_t, std::vector<std::unique_ptr<TfAllocatorAdapter>>>
+      memory_space_to_per_device_allocators_;
+  // Map of buffer to which memory space it resides in.
+  absl::Mutex mu_;
+  absl::flat_hash_map<void *, int64_t> buffer_memory_spaces_
+      ABSL_GUARDED_BY(mu);
```
Comment (cheshire) [inline xla/stream_executor/integrations/tf_allocator_adapter.h:178]: Could SE changes go into a separate PR?

**Review by cheshire — CHANGES_REQUESTED** [review]
Could we add tests?

**Comment by trevor-m** [comment]
Thank you @cheshire for the review! I will split the changes into 3 separate PRs:
Does this sound good to you?
1. xla/service/gpu Changes (will stay in this PR)
2. PJRT Changes
3. Stream Executor Changes

File: xla/stream_executor/integrations/device_mem_allocator.h (hunk 8 above)
Comment (trevor-m) [reply xla/stream_executor/integrations/device_mem_allocator.h:57]: This PR doesn't change anything in the way unified memory is handled. The change here is to the interface of `DeviceMemAllocator`. Previously, the allocator used a bool to determine whether to use one of two options: regular cuda memory or unified memory. This PR is adding a third option to use `ncclMemAlloc` memory, so I changed the bool to an enum.

File: xla/service/gpu/gpu_executable.cc (hunk 4 above)
Comment (trevor-m) [reply xla/service/gpu/gpu_executable.cc:835]: Yes that's correct.

The `colorerer` for BufferAssignments sets the memory space values for the allocations.
For the thunk runtime, those allocations are directly copied over to the executable and the memory space values are preserved.
However, for the non-thunk runtime, the allocations are reconstructed during `GetMlirAllocationInfo` and the memory space value gets erroneously set to 0 at this line.

File: xla/service/gpu/compile_module_to_llvm_ir.cc (hunk 2 above)
Comment (trevor-m) [reply xla/service/gpu/compile_module_to_llvm_ir.cc:315]: I will combine these.

File: xla/service/gpu/nccl_collective_thunk.cc (hunk 5 above)
Comment (trevor-m) [reply xla/service/gpu/nccl_collective_thunk.cc:230]: `ncclCommRegister` is a no-op if the buffer doesn't meet the requirements that `ncclMemAlloc` provides. However, I think I would still like to plumb the `xla_gpu_enable_nccl_user_buffers` flag through.

File: xla/pjrt/gpu/gpu_helpers.h [hunk 11]
```diff
@@ -57,6 +57,10 @@ struct GpuAllocatorConfig {
   // fragmentation, allowing more of the total memory to be used. If false, the
   // allocator will allocate more memory as allocations are requested.
   bool preallocate = true;
+
+  // Amount of collective memory (ncclMemAlloc) to reserve. If this value is 0,
+  // collective memory will not be used.
+  size_t collective_memory_size = 1 * (1LL << 30);
```
Comment (PatriosTheGreat) [inline xla/pjrt/gpu/gpu_helpers.h:63]: Correct me if I'm wrong should this size be the same as NCCL_BUFFSIZE?
If user will set different value here wouldn't it led to runtime failure?
If so should this parameter be boolean and could we get actual size from env variable?

File: xla/stream_executor/integrations/tf_allocator_adapter.h (hunk 10 above)
Comment (PatriosTheGreat) [inline xla/stream_executor/integrations/tf_allocator_adapter.h:178]: Should this be ABSL_GUARDED_BY(mu_); ?

File: xla/stream_executor/stream_executor_internal.h [hunk 12]
```diff
@@ -332,6 +332,8 @@ class StreamExecutorInterface {
   // Deallocates unified memory space previously allocated with
   // UnifiedMemoryAllocate.
   virtual void UnifiedMemoryDeallocate(void* mem) {}
+  virtual void* CollectiveMemoryAllocate(uint64_t size) = 0;
```
Comment (PatriosTheGreat) [inline xla/stream_executor/stream_executor_internal.h:335]: These methods should be implemented at the host_gpu_executor:
https://github.com/openxla/xla/blob/main/xla/stream_executor/host/host_gpu_executor.h#L66

Otherwise HostExecutor will became abstract class which will lead to compilation error (allocating an object of abstract class type) here:
https://github.com/openxla/xla/blob/main/xla/stream_executor/host/host_platform.cc#L63

File: xla/service/gpu/nccl_collective_thunk.cc [hunk 13]
```diff
@@ -224,6 +224,49 @@ StatusOr<std::vector<DeviceBufferPair>> ConvertToDeviceBuffers(
...
+    absl::Mutex mu;
+    absl::flat_hash_map<int, absl::flat_hash_set<ncclComm_t>> per_device_comms
+        ABSL_GUARDED_BY(mu);
+    // Buffers could be deregistered with ncclCommDeregister.
+    std::vector<void*> handles ABSL_GUARDED_BY(mu);
+  };
+  static auto& all_registered = *new RegisteredBuffers;
+
+  auto maybe_register = [&](const se::DeviceMemoryBase& buffer) {
+    if (!all_registered.per_device_comms[device_ordinal].contains(comm)) {
```
Comment (PatriosTheGreat) [inline xla/service/gpu/nccl_collective_thunk.cc:245]: Thread Safety Analysis at clang can't determine that this lambda is called only with mutex locked. So compilation with clang failing here with error:
error: reading variable 'per_device_comms' requires holding mutex 'all_registered.mu' [-Werror,-Wthread-safety-analysis]

There is an attribute which can disable this check for method:
https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#no-thread-safety-analysis

But unfortunately it doesn't work for lambdas. Can this lambda be extracted to different method if RegisteredBuffers will became singleton?

**Comment by cheshire** [comment]
>  I will split the changes into 3 separate PRs:  Does this sound good to you?

Yes, makes sense.

File: xla/pjrt/gpu/gpu_helpers.h (hunk 11 above)
Comment (trevor-m) [reply xla/pjrt/gpu/gpu_helpers.h:63]: This size we need here is determined by the peak memory usage of NCCL collective buffers in the model which is calculated during buffer assignment.

For example, an 8-gpu GPT5B needs around 400mb from looking at the dumped after-optimizations-buffer-assignment.txt file:
```
allocation 318: size 428487776, color 1, preallocated-temp:
 value: <1000 reduce-scatter.15 @1> (size=16777216,offset=896): bf16[16384,512]{0,1:S(1)}
 value: <1004 reduce-scatter.16 @1> (size=12582912,offset=33555328): bf16[3,32,128,512]{2,1,0,3:S(1)}
 . . .
 value: <4489 all-gather-start.10{1} @1> (size=411705344,offset=896): bf16[50257,4096]{0,1:S(1)}
```

Also, `ncclMemAlloc` aligns the size to the nearest 512mb so this number should be a multiple of 512mb to avoid wasting memory.

File: xla/service/gpu/nccl_collective_thunk.cc [hunk 14]
```diff
@@ -224,6 +224,49 @@ StatusOr<std::vector<DeviceBufferPair>> ConvertToDeviceBuffers(
...
+  // Each device has one NCCL buffer which only needs to be registered once per
+  // each comm.
+  struct RegisteredBuffers {
+    absl::Mutex mu;
+    absl::flat_hash_map<int, absl::flat_hash_set<ncclComm_t>> per_device_comms
+        ABSL_GUARDED_BY(mu);
+    // Buffers could be deregistered with ncclCommDeregister.
+    std::vector<void*> handles ABSL_GUARDED_BY(mu);
+  };
+  static auto& all_registered = *new RegisteredBuffers;
```
[the remark this answers is not in the document]
Comment (trevor-m) [reply xla/service/gpu/nccl_collective_thunk.cc:242]: For this code, I was mimicking how `AcquireNcclComm` stores the communicators: https://github.com/openxla/xla/blob/main/xla/service/gpu/nccl_utils.cc#L278

Although if this is not a good pattern to replicate, I can change this to a singleton.

**Comment by trevor-m** [comment]
I've split the changes into the following PRs:
PJRT Changes: https://github.com/openxla/xla/pull/7963
StreamExecutor Changes: https://github.com/openxla/xla/pull/7962
This PR contains the XLA changes.

**Review by PatriosTheGreat — APPROVED** [review]

File: xla/service/gpu/nccl_collective_thunk.cc (hunk 5 above)
Comment (PatriosTheGreat) [reply xla/service/gpu/nccl_collective_thunk.cc:230]: I assume it's still safer to check if user buffer is used here.
Can you pass the bool flag from debug_options through RunAllReduce/RunAllGather/OtherOperation methods?
Here: https://github.com/openxla/xla/blob/main/xla/service/gpu/runtime/collectives.cc#L553
And Here: https://github.com/openxla/xla/blob/main/xla/service/gpu/runtime/collectives.cc#L478
Etc?

@cheshire WDYT?

**Review by PatriosTheGreat — CHANGES_REQUESTED** [review]

**Comment by trevor-m** [comment]
> I assume it's still safer to check if user buffer is used here.
> Can you pass the bool flag from debug_options through RunAllReduce/RunAllGather/OtherOperation methods?
> Here: https://github.com/openxla/xla/blob/main/xla/service/gpu/runtime/collectives.cc#L553
> And Here: https://github.com/openxla/xla/blob/main/xla/service/gpu/runtime/collectives.cc#L478
> Etc?
> 
> @cheshire WDYT?

@PatriosTheGreat I've found a way to query if the pointer was allocated using `ncclMemAlloc`. I will change the code to only register the buffers if that's true. This is necessary since I've noticed sometimes not all collectives will use the nccl buffer and that would cause an error.

It will be great once NCCL collectives + cuda graphs are supported in XLA because buffer registration will no longer be required.

File: xla/service/gpu/nccl_utils.h [hunk 15]
```diff
@@ -41,6 +41,8 @@ limitations under the License.
 #include "rocm/include/rccl.h"
 #endif
 #else
+#include "third_party/gpus/cuda/include/cuda.h"
+#include "third_party/gpus/cuda/include/driver_types.h"
```
Comment (PatriosTheGreat) [inline xla/service/gpu/nccl_utils.h:45]: Shouldn't we add dependencies to these targets in bazel?

**Review by PatriosTheGreat — APPROVED** [review]
Thanks!

**Review by PatriosTheGreat — APPROVED** [review]

**Review by PatriosTheGreat — APPROVED** [review]
